# Feature Engineering Exercise (Core)

## Learning Objectives

- Use data to engineer new features to improve a model's ability to predict.

## Assignment

In this exercise, you will be working with data about bike share rentals. You can download the data here: https://docs.google.com/spreadsheets/d/e/2PACX-1vROUXPkYUkX-2W7JbJ0-oNKaXzpg4NtmU9IeWEY6yFKm32ZEJOpRh_soHD4BeIcuHjYik3SEoXmkgwj/pub?output=csv

Your task is to engineer some new features to try to improve a model's ability to predict the total number of bike share rentals during a given hour of the day.

## Step 1

1. Import the data the drop the 'casual' and 'registered' columns. These are redundant with your target, 'count'.


In [193]:
# import libraries
import pandas as pd

In [194]:
# load data
df = pd.read_csv('https://docs.google.com/spreadsheets/d/e/2PACX-1vROUXPkYUkX-2W7JbJ0-oNKaXzpg4NtmU9IeWEY6yFKm32ZEJOpRh_soHD4BeIcuHjYik3SEoXmkgwj/pub?output=csv')

# check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   datetime    10886 non-null  object 
 1   season      10886 non-null  int64  
 2   holiday     10886 non-null  int64  
 3   workingday  10886 non-null  int64  
 4   weather     10886 non-null  int64  
 5   temp        10886 non-null  float64
 6   atemp       10886 non-null  float64
 7   humidity    10886 non-null  int64  
 8   windspeed   10886 non-null  float64
 9   casual      10886 non-null  int64  
 10  registered  10886 non-null  int64  
 11  count       10886 non-null  int64  
dtypes: float64(3), int64(8), object(1)
memory usage: 1020.7+ KB


In [195]:
# check for duplicates
df.duplicated().sum()

0

In [196]:
# check for missing values
df.isna().sum().sum()

0

In [197]:
# check target column
df['count'].describe()

count    10886.000000
mean       191.574132
std        181.144454
min          1.000000
25%         42.000000
50%        145.000000
75%        284.000000
max        977.000000
Name: count, dtype: float64

In [198]:
# per the instructions, drop 'casual' and 'registered' columns
df.drop(columns = ['casual', 'registered'], inplace = True)

# check
df.columns

Index(['datetime', 'season', 'holiday', 'workingday', 'weather', 'temp',
       'atemp', 'humidity', 'windspeed', 'count'],
      dtype='object')

In [199]:
# check numeric columns
df.describe()

season       holiday    workingday       weather         temp  \
count  10886.000000  10886.000000  10886.000000  10886.000000  10886.00000   
mean       2.506614      0.028569      0.680875      1.418427     20.23086   
std        1.116174      0.166599      0.466159      0.633839      7.79159   
min        1.000000      0.000000      0.000000      1.000000      0.82000   
25%        2.000000      0.000000      0.000000      1.000000     13.94000   
50%        3.000000      0.000000      1.000000      1.000000     20.50000   
75%        4.000000      0.000000      1.000000      2.000000     26.24000   
max        4.000000      1.000000      1.000000      4.000000     41.00000   

              atemp      humidity     windspeed         count  
count  10886.000000  10886.000000  10886.000000  10886.000000  
mean      23.655084     61.886460     12.799395    191.574132  
std        8.474601     19.245033      8.164537    181.144454  
min        0.760000      0.000000      0.000000      1.000000  
25%       16.665000     47.000000      7.001500     42.000000  
50%       24.240000     62.000000     12.998000    145.000000  
75%       31.060000     77.000000     16.997900    284.000000  
max       45.455000    100.000000     56.996900    977.000000

## Step 2

2. Transform the 'datetime' column into a datetime type and use it to create 3 new columns in the data frame containing the:
  1. Name of the Month
  2. Name of the Day of the Week
  3. Hour of the Day
    1. Make sure all 3 new columns are 'object' datatype so they can be one-hot encoded later.
    2. Drop the 'datetime' and 'season' columns. These are now redundant.


In [200]:
# make a copy of df to save the original features
df2 = df.copy()

In [201]:
# change 'datetime' column to datetime dtype
df['datetime'] = pd.to_datetime(df['datetime'])

# check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   datetime    10886 non-null  datetime64[ns]
 1   season      10886 non-null  int64         
 2   holiday     10886 non-null  int64         
 3   workingday  10886 non-null  int64         
 4   weather     10886 non-null  int64         
 5   temp        10886 non-null  float64       
 6   atemp       10886 non-null  float64       
 7   humidity    10886 non-null  int64         
 8   windspeed   10886 non-null  float64       
 9   count       10886 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(6)
memory usage: 850.6 KB


In [202]:
# make new columns from 'datetime'
df['name_of_month'] = df['datetime'].dt.month_name()
df['name_of_weekday'] = df['datetime'].dt.day_name()
df['hour_of_day'] = df['datetime'].dt.hour

# check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   datetime         10886 non-null  datetime64[ns]
 1   season           10886 non-null  int64         
 2   holiday          10886 non-null  int64         
 3   workingday       10886 non-null  int64         
 4   weather          10886 non-null  int64         
 5   temp             10886 non-null  float64       
 6   atemp            10886 non-null  float64       
 7   humidity         10886 non-null  int64         
 8   windspeed        10886 non-null  float64       
 9   count            10886 non-null  int64         
 10  name_of_month    10886 non-null  object        
 11  name_of_weekday  10886 non-null  object        
 12  hour_of_day      10886 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(7), object(2)
memory usage: 1.1+ MB


In [203]:
# name_of_month and name_of_weekday are already objects,
# but hour_of_day should also be object for one-hot encoding
df['hour_of_day'] = df['hour_of_day'].astype(object)

# check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   datetime         10886 non-null  datetime64[ns]
 1   season           10886 non-null  int64         
 2   holiday          10886 non-null  int64         
 3   workingday       10886 non-null  int64         
 4   weather          10886 non-null  int64         
 5   temp             10886 non-null  float64       
 6   atemp            10886 non-null  float64       
 7   humidity         10886 non-null  int64         
 8   windspeed        10886 non-null  float64       
 9   count            10886 non-null  int64         
 10  name_of_month    10886 non-null  object        
 11  name_of_weekday  10886 non-null  object        
 12  hour_of_day      10886 non-null  object        
dtypes: datetime64[ns](1), float64(3), int64(6), object(3)
memory usage: 1.1+ MB


In [204]:
# per instructions, drop 'datetime' and 'season' columns
df.drop(columns = ['datetime', 'season'], inplace = True)

# check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   holiday          10886 non-null  int64  
 1   workingday       10886 non-null  int64  
 2   weather          10886 non-null  int64  
 3   temp             10886 non-null  float64
 4   atemp            10886 non-null  float64
 5   humidity         10886 non-null  int64  
 6   windspeed        10886 non-null  float64
 7   count            10886 non-null  int64  
 8   name_of_month    10886 non-null  object 
 9   name_of_weekday  10886 non-null  object 
 10  hour_of_day      10886 non-null  object 
dtypes: float64(3), int64(5), object(3)
memory usage: 935.6+ KB


## Step 3

3. The temperatures in the 'temp' and 'atemp' columns are in Celsius. Use `.apply()` and a Lambda function to convert them to Fahrenheit.

In [205]:
# convert 'temp' and 'atemp' to fahrenheit
df[['f_temp', 'f_atemp']] = df[['temp', 'atemp']].apply(lambda x: ((x * (9/5)) + 32))

# check
display(df[['temp', 'atemp']])
display(df[['f_temp', 'f_atemp']])

temp   atemp
0       9.84  14.395
1       9.02  13.635
2       9.02  13.635
3       9.84  14.395
4       9.84  14.395
...      ...     ...
10881  15.58  19.695
10882  14.76  17.425
10883  13.94  15.910
10884  13.94  17.425
10885  13.12  16.665

[10886 rows x 2 columns]

f_temp  f_atemp
0      49.712   57.911
1      48.236   56.543
2      48.236   56.543
3      49.712   57.911
4      49.712   57.911
...       ...      ...
10881  60.044   67.451
10882  58.568   63.365
10883  57.092   60.638
10884  57.092   63.365
10885  55.616   61.997

[10886 rows x 2 columns]

In [206]:
# drop 'temp' and 'atemp' columns
df.drop(columns = ['temp', 'atemp'], inplace = True)

# check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   holiday          10886 non-null  int64  
 1   workingday       10886 non-null  int64  
 2   weather          10886 non-null  int64  
 3   humidity         10886 non-null  int64  
 4   windspeed        10886 non-null  float64
 5   count            10886 non-null  int64  
 6   name_of_month    10886 non-null  object 
 7   name_of_weekday  10886 non-null  object 
 8   hour_of_day      10886 non-null  object 
 9   f_temp           10886 non-null  float64
 10  f_atemp          10886 non-null  float64
dtypes: float64(3), int64(5), object(3)
memory usage: 935.6+ KB


## Step 4

4. Create a new column, 'temp_variance' which shows how much warmer or colder the current temperature ('temp') is than the average temperate for that day of the year ('atemp').  If the current temperature is warmer than average ('atemp'), the value in 'temp_variance' should be positive. 
  1. Drop the 'atemp' column.

In [207]:
# create 'temp_variance'
df['temp_variance'] = df['f_temp'] - df['f_atemp']

# check
df[['f_temp', 'f_atemp', 'temp_variance']]

f_temp  f_atemp  temp_variance
0      49.712   57.911         -8.199
1      48.236   56.543         -8.307
2      48.236   56.543         -8.307
3      49.712   57.911         -8.199
4      49.712   57.911         -8.199
...       ...      ...            ...
10881  60.044   67.451         -7.407
10882  58.568   63.365         -4.797
10883  57.092   60.638         -3.546
10884  57.092   63.365         -6.273
10885  55.616   61.997         -6.381

[10886 rows x 3 columns]

In [208]:
# drop 'f_atemp'
df.drop(columns = 'f_atemp', inplace = True)

# check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   holiday          10886 non-null  int64  
 1   workingday       10886 non-null  int64  
 2   weather          10886 non-null  int64  
 3   humidity         10886 non-null  int64  
 4   windspeed        10886 non-null  float64
 5   count            10886 non-null  int64  
 6   name_of_month    10886 non-null  object 
 7   name_of_weekday  10886 non-null  object 
 8   hour_of_day      10886 non-null  object 
 9   f_temp           10886 non-null  float64
 10  temp_variance    10886 non-null  float64
dtypes: float64(3), int64(5), object(3)
memory usage: 935.6+ KB


## Optional

Use a predictive model of your choice and try to predict the 'count' of hourly bike-share users with both the original features and the engineered feature set you created.

Remember to drop the 'casual' and 'registered' columns from both versions before modeling.

Did these feature engineering choices improve your ability to predict the 'count'?

In [209]:
# split data into X and y to prep for supervised machine learning
target = 'count'

# split for original features
y2 = df2[target]
X2 = df2.drop(columns = target)

# split for engineered features
y1 = df[target]
X1 = df.drop(columns = target)

In [210]:
# train/test split for model validation
from sklearn.model_selection import train_test_split
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, random_state = 42)
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, random_state = 42)

### Original Features

In [211]:
# check original features in df2
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   datetime    10886 non-null  object 
 1   season      10886 non-null  int64  
 2   holiday     10886 non-null  int64  
 3   workingday  10886 non-null  int64  
 4   weather     10886 non-null  int64  
 5   temp        10886 non-null  float64
 6   atemp       10886 non-null  float64
 7   humidity    10886 non-null  int64  
 8   windspeed   10886 non-null  float64
 9   count       10886 non-null  int64  
dtypes: float64(3), int64(6), object(1)
memory usage: 850.6+ KB


In [212]:
# let's try a simple linear regression for this problem

# model 1: linear regression without feature engineering

# check out number of unique values in 'datetime'
df2['datetime'].nunique()

10886

In [213]:
# there are as many unique values in 'datetime' as there are
# rows, so we will drop this column from model 1

df2.drop(columns = 'datetime', inplace = True)
X2.drop(columns = 'datetime', inplace = True)
X2_train.drop(columns = 'datetime', inplace = True)
X2_test.drop(columns = 'datetime', inplace = True)

# check
df2.info()
X2.info()
X2_train.info()
X2_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   season      10886 non-null  int64  
 1   holiday     10886 non-null  int64  
 2   workingday  10886 non-null  int64  
 3   weather     10886 non-null  int64  
 4   temp        10886 non-null  float64
 5   atemp       10886 non-null  float64
 6   humidity    10886 non-null  int64  
 7   windspeed   10886 non-null  float64
 8   count       10886 non-null  int64  
dtypes: float64(3), int64(6)
memory usage: 765.5 KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   season      10886 non-null  int64  
 1   holiday     10886 non-null  int64  
 2   workingday  10886 non-null  int64  
 3   weather     10886 non-null  int64  
 4   temp        10886 non-null  fl

In [214]:
# preprocessing plan for original features
# season, weather: ohe
# holiday, workingday: passthrough (binary)
# temp, atemp, humidity, windspeed: scale

from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import make_column_transformer

# make ohe tuple
ohe_tuple = (OneHotEncoder(), ['season', 'weather'])

# make scaler tuple
scaler_tuple = (StandardScaler(), ['temp', 'atemp', 'humidity', 'windspeed'])

# make preprocessor
preprocessor2 = make_column_transformer(ohe_tuple,
                                       scaler_tuple,
                                       remainder = 'passthrough')

# check
preprocessor2

ColumnTransformer(remainder='passthrough',
                  transformers=[('onehotencoder', OneHotEncoder(),
                                 ['season', 'weather']),
                                ('standardscaler', StandardScaler(),
                                 ['temp', 'atemp', 'humidity', 'windspeed'])])

In [215]:
# instantiate original_features_pipe

from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LinearRegression

original_features_pipe = make_pipeline(preprocessor2, LinearRegression())

# check
original_features_pipe

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(),
                                                  ['season', 'weather']),
                                                 ('standardscaler',
                                                  StandardScaler(),
                                                  ['temp', 'atemp', 'humidity',
                                                   'windspeed'])])),
                ('linearregression', LinearRegression())])

In [216]:
# fit on training data
original_features_pipe.fit(X2_train, y2_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(),
                                                  ['season', 'weather']),
                                                 ('standardscaler',
                                                  StandardScaler(),
                                                  ['temp', 'atemp', 'humidity',
                                                   'windspeed'])])),
                ('linearregression', LinearRegression())])

In [217]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
import numpy as np

# evaluate on testing data
y2_test_preds = original_features_pipe.predict(X2_test)

print(f"r2: {r2_score(y2_test, y2_test_preds)}")
print(f"mae: {mean_absolute_error(y2_test, y2_test_preds)}")
print(f"rmse: {np.sqrt(mean_squared_error(y2_test, y2_test_preds))}")

r2: 0.2788854835114559
mae: 113.79821822189567
rmse: 153.82678784140526


### Engineered Features

In [145]:
# check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   holiday          10886 non-null  int64  
 1   workingday       10886 non-null  int64  
 2   weather          10886 non-null  int64  
 3   humidity         10886 non-null  int64  
 4   windspeed        10886 non-null  float64
 5   count            10886 non-null  int64  
 6   name_of_month    10886 non-null  object 
 7   name_of_weekday  10886 non-null  object 
 8   hour_of_day      10886 non-null  object 
 9   f_temp           10886 non-null  float64
 10  temp_variance    10886 non-null  float64
dtypes: float64(3), int64(5), object(3)
memory usage: 935.6+ KB


In [218]:
# preprocessing plan for engineered features
# weather, name_of_month, name_of_weekday, hour_of_day: ohe
# holiday, workingday: passthrough (binary)
# f_temp, temp_variance, humidity, windspeed: scale

# make ohe tuple
ohe_tuple = (OneHotEncoder(), ['weather', 
                               'name_of_month', 
                               'name_of_weekday', 
                               'hour_of_day'])

# make scaler tuple
scaler_tuple = (StandardScaler(), ['f_temp', 
                                   'temp_variance', 
                                   'humidity', 
                                   'windspeed'])

# make preprocessor
preprocessor1 = make_column_transformer(ohe_tuple,
                                       scaler_tuple,
                                       remainder = 'passthrough')

# check
preprocessor1

ColumnTransformer(remainder='passthrough',
                  transformers=[('onehotencoder', OneHotEncoder(),
                                 ['weather', 'name_of_month', 'name_of_weekday',
                                  'hour_of_day']),
                                ('standardscaler', StandardScaler(),
                                 ['f_temp', 'temp_variance', 'humidity',
                                  'windspeed'])])

In [219]:
# instantiate engineered_features_pipe

engineered_features_pipe = make_pipeline(preprocessor1, LinearRegression())

# check
engineered_features_pipe

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(),
                                                  ['weather', 'name_of_month',
                                                   'name_of_weekday',
                                                   'hour_of_day']),
                                                 ('standardscaler',
                                                  StandardScaler(),
                                                  ['f_temp', 'temp_variance',
                                                   'humidity',
                                                   'windspeed'])])),
                ('linearregression', LinearRegression())])

In [220]:
# fit on training data
engineered_features_pipe.fit(X1_train, y1_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(),
                                                  ['weather', 'name_of_month',
                                                   'name_of_weekday',
                                                   'hour_of_day']),
                                                 ('standardscaler',
                                                  StandardScaler(),
                                                  ['f_temp', 'temp_variance',
                                                   'humidity',
                                                   'windspeed'])])),
                ('linearregression', LinearRegression())])

In [221]:
# evaluate on testing data
y1_test_preds = engineered_features_pipe.predict(X1_test)

print(f"r2: {r2_score(y1_test, y1_test_preds)}")
print(f"mae: {mean_absolute_error(y1_test, y1_test_preds)}")
print(f"rmse: {np.sqrt(mean_squared_error(y1_test, y1_test_preds))}")

r2: 0.6287021620967724
mae: 80.23781247688166
rmse: 110.38017095648003


### Comparison Results

The linear regression model with engineered features produced a much higher R2 score than the one with original features (0.63 to 0.28). The MAE and RMSE scores were also lower for the engineered features model.